### DISCUSSION TITLE
Invalid configuration for device '0'

### DISCUSSION THREAD

**Discussion by kmcgrego1307** [vmware/packer-examples-for-vsphere#163] [Help] [answered]
I tried a build of Ubuntu Server 20.04, and got this error. I don't know what I should be looking at to fix this. The ISO is there, and I updated the SHA256 checksum in the linux-ubuntu.auto.pkrvars.hcl file. Can someone help?

CONFIRM: Build a Ubuntu Server 20.04 LTS Template for VMware vSphere?

Continue? (y/n)
y
Building a Ubuntu Server 20.04 LTS Template for VMware vSphere...
Initializing HashiCorp Packer and required plugins...
Starting the build....
vsphere-iso.linux-ubuntu: output will be in this color.

==> vsphere-iso.linux-ubuntu: Creating VM...
==> vsphere-iso.linux-ubuntu: Customizing hardware...
==> vsphere-iso.linux-ubuntu: Adding SATA controller...
==> vsphere-iso.linux-ubuntu: Mounting ISO images...
==> vsphere-iso.linux-ubuntu: Destroying VM...
Build 'vsphere-iso.linux-ubuntu' errored after 1 second 732 milliseconds: error mounting an image '[<redacted>] ISOs/Linux/ubuntu-20.04.4-live-server-amd64.iso': Invalid configuration for device '0'.

==> Wait completed after 1 second 732 milliseconds

==> Some builds didn't complete successfully and had errors:
--> vsphere-iso.linux-ubuntu: error mounting an image '[<redacted>] ISOs/Linux/ubuntu-20.04.4-live-server-amd64.iso': Invalid configuration for device '0'.

==> Builds finished but no artifacts were created.

**Comment by tenthirtyam** [comment, marked as answer]
In a previous message you mentioned that your environment uses vSphere 6.7 U2.
The project defaults to virtual hardware 19, and I suspect you need to modify the `common_vm_version` in the `common.pkvars.hcl` down to a version (15) compatible with your environment.

See https://kb.vmware.com/s/article/1003746.

Ryan

**Reply by kmcgrego1307** [reply]
I had previously changed that setting from "19" to "14". This produced the error above. I also tried "15" and "11", and I get the same error message.

**Reply by tenthirtyam** [reply]
Any other significant changes from the defaults?

**Reply by kmcgrego1307** [reply]
No. I changed as little as possible to get even this far. Mainly the obvious settings that had to change to reflect our environment (e.g. names of things).
I just noticed the Recent Tasks in the vSphere Client shows
Reconfigure virtual machine   linux-ubuntu-20.04lts-v22.03   Invalid configuration for device '0'. Device: VirtualCdrom

I did set the vm_firmware to 'bios', though since you said the firmware would revert to that anyway. 'efi' gives the same error.

**Reply by tenthirtyam** [reply]
EFI is **required** for the machine images as the kickstart and partitioning is set for use of EFI.

In the pkvars.hcl for Ubuntu, change the 
`vm_cdrom_type` from "sata" to "ide" and give that a chance.

Ryan

**Reply by kmcgrego1307** [reply]
I tried vm_firmware = "bios" with vm_cdrom_type = "ide" and it still fails with the same error.

Side note: I asked if/when we're upgrading to ESXi 7.x and I was told that would take place this summer. Since one of our biggest (internal) customers is planning a Ubuntu upgrade to 22.04 in April, I hoped I could get this working before the vSphere upgrade happens.

**Reply by tenthirtyam** [reply]
Using `vm_firmware = "bios"` will certainly break the kickstart and will not work with the provided disk layout. Both can be changed but would require changes to command keys and the kickstart template. 

> Note: You should ensure you review the kickstart file example and refine for your requirements.

I don't presently have a 6.7 testbed but if I have time to engage one, set the requirements, and run some unit tests I'll let you know.

Also, I'll be adding Ubuntu 22.04 LTS in April after its GA.

Ryan

**Reply by tenthirtyam** [reply]
```shell
--> vsphere-iso.linux-ubuntu: error mounting an image '[] ISOs/Linux/ubuntu-20.04.4-live-server-amd64.iso': Invalid configuration for device '0'.
```

Looks like it could be the paths to your ISO. It may be failing to mount file and connect.

It's failing in the "Mounting ISOs..." stage.

Ryan

**Reply by kmcgrego1307** [reply]
It turns out I had selected an inappropriate datastore for the ISO. I found one that works.
